### PR TITLE
Render the controller name in the model table

### PR DESCRIPTION
**Pull request by anthonydillon**
## Done
Render the controller name in the model table

## QA
- `juju bootstrap` into a public cloud
- `juju dashboard` and connect to that dashboard to accept the self-signed cert.
  - take note of the ip address and u/p
- Follow the deployment steps in #526 
- Open config.js and change: 
  - `baseControllerURL` to the ip address and port (if any) from above.
  - `isJuju` to `true`
  - `identityProviderAvailable` to `false`
- `./run`
- Open http://0.0.0.0:8036/
- You should see your controller name in the list instead of the uuid

## Details
Fixes https://github.com/canonical-web-and-design/jaas-dashboard/issues/532

## Screenshots
image

### CODE REVIEW

File: src/components/ModelTableList/shared.js [hunk 1]
```diff
@@ -136,21 +139,3 @@ export function getStatusValue(status, key) {
...
   return returnValue;
 }
-
-/**
-  Renders the controller UUID element. If the UUID is the JAAS controller then
-  it replaces the UUID with "JAAS".
-  @param {string} uuid The controller UUID.
-  @returns {Object} The element to render the controller UUID entry
-*/
-export function generateControllerUUID(uuid) {
```
Comment (anthonydillon) [inline src/components/ModelTableList/shared.js:146]: Removed this function as its not used anywhere anymore.

**Comment by anthonydillon** [comment]
@hatched @barrymcgee this is now ready with tests.

File: src/app/selectors.js [hunk 2]
```diff
@@ -367,9 +370,35 @@ const countModelStatusGroups = (groupedModelStatuses) => {
...
-const filterModelData = (filters, modelData) => {
+const filterModelData = (filters, modelData, controllerData) => {
+  const clonedModelData = cloneDeep(modelData);
+  // Add controllerName to the model data by mapping the controllerData
+  Object.entries(clonedModelData).forEach((model) => {
+    if (model[1].info) {
+      let controllerName = null;
+      if (controllerData) {
+        for (var i = 0; i < controllerData.length; i++) {
+          if (controllerData[i].uuid === model[1].info.controllerUuid) {
```
Comment (hatched) [inline src/app/selectors.js:381]: a bit pedantic but can you invert this....you're actually checking if the controller uuid is the same one in the model....so the model goes first 😆

File: src/app/selectors.js [hunk 3]
```diff
@@ -367,9 +370,35 @@ const countModelStatusGroups = (groupedModelStatuses) => {
   @param {Object} modelData The model data from the redux store.
   @returns {Object} The filtered model data.
 */
-const filterModelData = (filters, modelData) => {
+const filterModelData = (filters, modelData, controllerData) => {
+  const clonedModelData = cloneDeep(modelData);
+  // Add controllerName to the model data by mapping the controllerData
+  Object.entries(clonedModelData).forEach((model) => {
+    if (model[1].info) {
+      let controllerName = null;
```
Comment (hatched) [inline src/app/selectors.js:378]: If you set the default value to the UUID then you can remove the `else` part of the conditional on 392.
```suggestion
      let controllerName = model[1].info.controllerUuid;
```

File: src/components/ModelTableList/CloudGroup.js [hunk 4]
```diff
@@ -67,48 +63,57 @@ export default function CloudGroup({ activeUser, filters }) {
       modelGroup.forEach((model) => {
         const { highestStatus } = getModelStatusGroupData(model);
         cloudModels.rows.push({
+          "data-test-model-id": model?.uuid,
```
Comment (hatched) [inline src/components/ModelTableList/CloudGroup.js:66]: Maybe we should be more specific in our `data-test` attributes, so 'id' to 'uuid' here? If you agree the other files will need to be updated as well.

```suggestion
          "data-test-model-uuid": model?.uuid,
```

File: src/components/ModelTableList/CloudGroup.js [hunk 5]
```diff
@@ -67,48 +63,57 @@ export default function CloudGroup({ activeUser, filters }) {
...
           columns: [
             {
+              "data-test-column": "name",
```
Comment (hatched) [inline src/components/ModelTableList/CloudGroup.js:69]: What do we think of using the prefix `data-test` for test-only attributes? Typically the attribute is simply `data-test` and then the value has the additional context to it.

I think I'm ok with this as it's still understandable and grep-able.

@barrymcgee

File: src/components/ModelTableList/CloudGroup.js [hunk 6]
```diff
@@ -67,48 +63,57 @@ export default function CloudGroup({ activeUser, filters }) {
...
               ),
             },
-            // We're not currently able to get the controller name from the API
-            // so, display the controller UUID instead.
             {
-              content: generateControllerUUID(
-                getStatusValue(model.info, "controllerUuid")
+              "data-test-column": "controller",
+              content: (
+                <span data-test="controller-name">
```
Comment (hatched) [inline src/components/ModelTableList/CloudGroup.js:109]: Is it necessary to wrap this in a span and add another data attribute?

File: src/components/ModelTableList/ModelTableList.test.js [hunk 7]
```diff
@@ -72,4 +73,57 @@ describe("ModelTableList", () => {
       expect(wrapper.find(table.component).prop("filters")).toBe(filters);
     });
   });
+
+  it("renders the controller name as JAAS", () => {
```
Comment (hatched) [inline src/components/ModelTableList/ModelTableList.test.js:77]: Great tests! 🎉

File: src/components/ModelTableList/CloudGroup.js (hunk 5 above)
Comment (anthonydillon) [reply src/components/ModelTableList/CloudGroup.js:69]: I must prefer having bespoke attributes as it causes a separation of concerns between functionality and testing. It makes it clear from the code what the purpose of the attribute.

It's absolutely normal to have multiple word data attributes: https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes

File: src/app/selectors.js [hunk 8]
```diff
@@ -367,9 +370,35 @@ const countModelStatusGroups = (groupedModelStatuses) => {
...
+      if (controllerData) {
+        for (var i = 0; i < controllerData.length; i++) {
```
[the remark this answers is not in the document]
Comment (anthonydillon) [reply src/app/selectors.js:380]: I know this is the new coolness but there is something nice about an old fashioned for loop with a `break;` :stuck_out_tongue:

File: src/components/ModelTableList/CloudGroup.js (hunk 4 above)
Comment (anthonydillon) [reply src/components/ModelTableList/CloudGroup.js:66]: Good idea.

File: src/app/selectors.js [hunk 9]
```diff
@@ -367,9 +370,35 @@ const countModelStatusGroups = (groupedModelStatuses) => {
   @param {Object} modelData The model data from the redux store.
   @returns {Object} The filtered model data.
 */
-const filterModelData = (filters, modelData) => {
+const filterModelData = (filters, modelData, controllerData) => {
+  const clonedModelData = cloneDeep(modelData);
+  // Add controllerName to the model data by mapping the controllerData
```
[the remark this answers is not in the document]
Comment (anthonydillon) [reply src/app/selectors.js:375]: Done

File: src/app/selectors.js (hunk 2 above)
Comment (anthonydillon) [reply src/app/selectors.js:381]: Done

File: src/app/selectors.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (anthonydillon) [reply src/app/selectors.js:380]: Done

File: src/app/selectors.js (hunk 3 above)
Comment (anthonydillon) [reply src/app/selectors.js:378]: Done

File: src/components/ModelTableList/CloudGroup.js (hunk 6 above)
Comment (anthonydillon) [reply src/components/ModelTableList/CloudGroup.js]: Not at all. Leftover from before I know arbitrary attributes can do add.

File: src/components/ModelTableList/CloudGroup.js (hunk 5 above)
Comment (hatched) [reply src/components/ModelTableList/CloudGroup.js:69]: Sounds all good to me!

**Comment by hatched** [comment]
QA Note:

Looks like the column is empty on JAAS
Screen Shot 2020-06-02 at 3 25 37 PM

**Comment by hatched** [comment]
But it works ok on a self-bootstrapped controller.

**Comment by hatched** [comment]
Thanks for the updates, looks good both on JAAS and Juju

**Pull request merged**